src/cljc/athens/common_events/graph/ops.cljc

```diff
@@ -195,3 +196,39 @@
                            (map :block/new)
                            set)]
     new-uids))
+
+
+(defn structural-diff
```

Comment:
nit: I think this function is mostly doing the same as linkmaker (especially the 1-arity), except it looks at the atomic op string instead of datoms.

Linkmaker 1-arity looks like this:
```clj
     (let [datoms        (d/datoms db :eavt)
           linkmaker-txs (into []
                               (comp (keep parseable-string-datom)
                                     (mapcat (partial update-refs db)))
                               datoms)]

```

It gets entity ids and strings from the db, and creates a tx via `update-refs`:

```clj
(defn update-refs-tx
  "Return the tx that will update lookup ref's :block/refs from before to after.
   Both before and after should be sets of lookup refs."
  [lookup-ref before after]
  (let [[only-before only-after] (data/diff before after)
        to-tx (fn [type ref] [type lookup-ref :block/refs ref])]
    (set (concat (map (partial to-tx :db/retract) only-before)
                 (map (partial to-tx :db/add) only-after)))))

(defn update-refs
  "Returns updated refs for eid.
   Returns nil if eid is no longer in db."
  [db [eid string]]
  (when-let [lookup-ref (eid->lookup-ref db eid)]
    (let [before     (block-refs-as-lookup-refs db lookup-ref)
          after      (string-as-lookup-refs db string)]
      (update-refs-tx lookup-ref before after))))
```

The overall logic here is the same I think: given a db, a entity, and a string, check the difference in refs between what's in the db right now and what's in the string. But on this function you don't want the output as a tx, I think you just want the `data/diff` between the lookups, because `data/diff` gives you `[removed-links added-links]`.

Reply:
I've tried `data/diff` it's output is confusing with all these `nil` all over the place.
Maybe linkmaker could use the structural-diff, though it's entangled with Atomic Ops.

I agree that similar things are happening in both places.
We don't have a common abstraction between the two just now.

Reply:
Then let's wait until the third case where we're doing the same similar thing to create the abstraction.